lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+          if (div->getArg() != 0) {
+            LOG(INFO) << div->getArg() << " : " << div->getName().data();
+            param_names[div->getArg()] = div->getName().data();
+          }
+        } else if (auto value_intrin =
+                       llvm::dyn_cast<llvm::DbgValueInst>(debug_inst)) {
+          const llvm::MDNode *mdn = value_intrin->getVariable();
+          const llvm::DILocalVariable *div =
+              llvm::cast<llvm::DILocalVariable>(mdn);
+
+          if (div->getArg() != 0) {
+            LOG(INFO) << div->getArg() << " : " << div->getName().data();
+            param_names[div->getArg()] = div->getName().data();
+          }
+        }
+      }
+    }
+  }
+
+  // If we don't have names for some parameters then automatically name them
+  unsigned int num_args =
+      (unsigned int)(function.args().end() - function.args().begin());
+  for (unsigned int i = 1; i <= num_args; i++) {
+    if (!param_names.count(i)) {
+      param_names[i] = "param" + std::to_string(i);
+    }
+  }
+
+  return param_names;
+}
+
+// TODO: This doesn't really fit here so I will need to find a better place for
+// this. Translates an llvm::Type to a type that conforms to the spec in
+// TypeParser.cpp
+std::string TranslateType(const llvm::Type &type) {
+  unsigned int id = type.getTypeID();
+
+  std::string ret;
+
+  // Types should match the type parsing in TypeParser.cpp
+  switch (id) {
+    case llvm::Type::VoidTyID: {
+      ret = "v";
+      break;
+    }
+    case llvm::Type::HalfTyID: {
+      ret = "e";
+      break;
+    }
+    case llvm::Type::FloatTyID: {
+      ret = "f";
+      break;
+    }
+    case llvm::Type::DoubleTyID: {
+      ret = "d";
+      break;
+    }
+    case llvm::Type::X86_FP80TyID: {
+      ret = "D";
+      break;
+    }
+    case llvm::Type::X86_MMXTyID: {
+      ret = "M";
+      break;
+    }
+    case llvm::Type::IntegerTyID: {
+      auto derived = llvm::cast<llvm::IntegerType>(type);
+      // Since there is no way to check for uint vs. int, lower all integer
+      // types to signed integers. Maybe come back when there is a better way to
+      // figure out unsigned vs signed ints
+      auto sign = true;
+      switch (derived.getBitWidth()) {
+        case 8: {
+          ret = sign ? "b" : "B";
+          break;
+        }
+        case 16: {
+          ret = sign ? "h" : "H";
+          break;
+        }
+        case 32: {
+          ret = sign ? "i" : "I";
+          break;
+        }
+        case 64: {
+          ret = sign ? "l" : "L";
+          break;
+        }
+      }
+      break;
+    }
+    case llvm::Type::FunctionTyID: {
+      ret = "func";
+      break;
+    }
+    case llvm::Type::StructTyID: {
+      auto struct_ptr = llvm::cast<llvm::StructType>(&type);
+      std::string element_list = "";
+      for (unsigned i = 0; i < struct_ptr->getNumElements(); i++) {
+        // TODO: BUG: this if statement protects against infinite chained
```

Comment:
This needs to be solved. The way to solve it is to pass a "cache" of types down as an argument to this function. We know that structs are the only such type than can enter into this situation. You can see this pattern used here, where `ids` is the cache: https://github.com/lifting-bits/anvill/blob/master/python/anvill/type.py#L240-L253